File: frontend/lib/trace/engine/svgWriter.ts
import { FillRule, LineCap, LineJoin, VecDocument, VecShape, VecStyle } from './path';
import { toD } from './svgPathData';

/**
 * SVG serialisation.
 *
 * Colours are written as `#rrggbb` with a separate `*-opacity` attribute whenever alpha is below 255.
 * `#rrggbbaa` is valid CSS Color 4 but is not understood by several cutting-plotter and CNC tool chains
 * that are a primary target of this app, and a silently ignored alpha there means a stroke that plots as
 * fully opaque.
 */

export interface SvgOptions {
  readonly precision: number;
  readonly includeMetadata: boolean;
  readonly groupByLayer: boolean;
  readonly prettyPrint: boolean;
  readonly widthUnit: string;
  readonly title: string;
}

export const DEFAULT_SVG_OPTIONS: SvgOptions = {
  precision: 2,
  includeMetadata: true,
  groupByLayer: true,
  prettyPrint: true,
  widthUnit: 'px',
  title: 'Offline Tracer export',
};

/** @returns {@link DEFAULT_SVG_OPTIONS} with `over` applied. */
export function svgOptions(over: Partial<SvgOptions> = {}): SvgOptions {
  return { ...DEFAULT_SVG_OPTIONS, ...over };
}

/** Escapes the five XML metacharacters. Applied to every string that reaches the output. */
export function escapeXml(s: string): string {
  let out = '';
  for (let i = 0; i < s.length; i++) {
    const c = s[i];
    if (c === '&') out += '&amp;';
    else if (c === '<') out += '&lt;';
    else if (c === '>') out += '&gt;';
    else if (c === '"') out += '&quot;';
    else if (c === "'") out += '&apos;';
    // XML 1.0 cannot represent the C0 controls other than tab, newline and carriage return at all —
    // not even as a numeric entity — so they are dropped rather than escaped. A layer name pasted
    // from a spreadsheet routinely carries one, and a single stray 0x01 makes the whole file
    // unopenable in every parser.
    else if (c >= ' ' || c === '\t' || c === '\n' || c === '\r') out += c;
  }
  return out;
}

function hex2(v: number): string {
  const s = (v & 0xff).toString(16);
  return s.length === 1 ? `0${s}` : s;
}

function colourHex(argb: number): string {
  return `#${hex2(argb >>> 16)}${hex2(argb >>> 8)}${hex2(argb)}`;
}

/**
 * Fixed-point, locale-independent, and **never non-finite**.
 *
 * A single `NaN` or `Infinity` anywhere in an SVG attribute makes the renderer drop the element that
 * carries it — silently, and only in some renderers — so a finite substitute is always emitted. A wrong
 * pixel is debuggable; a blank export is not.
 */
function num(v: number, precision: number): string {
  const finite = Number.isFinite(v) ? v : 0;
  const clamped = finite > MAX_COORD ? MAX_COORD : finite < -MAX_COORD ? -MAX_COORD : finite;
  let s = clamped.toFixed(precision);
  if (s.indexOf('.') >= 0) s = s.replace(/0+$/, '').replace(/\.$/, '');
  return s === '-0' ? '0' : s;
}

/** Beyond this a coordinate is meaningless and some renderers overflow their fixed-point maths. */
const MAX_COORD = 1e7;

/** A non-positive or non-finite canvas dimension falls back to 1 unit so the file still opens. */
function sanitizeDimension(v: number): number {
  if (!Number.isFinite(v) || v <= 0) return 1;
  return v > MAX_COORD ? MAX_COORD : v;
}

function capName(c: LineCap): string {
  if (c === LineCap.ROUND) return 'round';
  if (c === LineCap.SQUARE) return 'square';
  return 'butt';
}

function joinName(j: LineJoin): string {
  if (j === LineJoin.ROUND) return 'round';
  if (j === LineJoin.BEVEL) return 'bevel';
  return 'miter';
}

function styleAttrs(style: VecStyle, precision: number): string {
  const parts: string[] = [];
  if (style.fill === null) {
    parts.push('fill="none"');
  } else {
    parts.push(`fill="${colourHex(style.fill)}"`);
    const a = (style.fill >>> 24) & 0xff;
    if (a !== 255) parts.push(`fill-opacity="${num(a / 255, 3)}"`);
    parts.push(`fill-rule="${style.fillRule === FillRule.NONZERO ? 'nonzero' : 'evenodd'}"`);
  }
  if (style.stroke === null) {
    parts.push('stroke="none"');
  } else {
    parts.push(`stroke="${colourHex(style.stroke)}"`);
    const a = (style.stroke >>> 24) & 0xff;
    if (a !== 255) parts.push(`stroke-opacity="${num(a / 255, 3)}"`);
    parts.push(`stroke-width="${num(style.strokeWidth, precision)}"`);
    parts.push(`stroke-linecap="${capName(style.cap)}"`);
    parts.push(`stroke-linejoin="${joinName(style.join)}"`);
    if (style.join === LineJoin.MITER && style.miterLimit !== 4) {
      parts.push(`stroke-miterlimit="${num(style.miterLimit, precision)}"`);
    }
  }
  if (style.opacity !== 1) parts.push(`opacity="${num(style.opacity, 3)}"`);
  return parts.join(' ');
}

function shapeElement(shape: VecShape, options: SvgOptions, indent: string): string {
  const d = toD(shape.path, options.precision);
  const attrs = styleAttrs(shape.style, options.precision);
  const id = shape.path.id !== '' ? ` id="${escapeXml(shape.path.id)}"` : '';
  return `${indent}<path${id} ${attrs} d="${d}"/>`;
}

/**
 * Serialise a document.
 *
 * The `viewBox` is always `0 0 width height` — the document's own canvas, not the shapes' bounding box.
 * Cropping to the content would move the artwork relative to the source image, and a user who traces a
 * photo and then imports the SVG over it expects the two to line up.
 *
 * @returns a complete standalone SVG document with an XML declaration.
 */
export function write(doc: VecDocument, options: SvgOptions = DEFAULT_SVG_OPTIONS): string {
  const p = Math.max(0, Math.min(8, options.precision | 0));
  const opts: SvgOptions = { ...options, precision: p };
  const docW = sanitizeDimension(doc.width);
  const docH = sanitizeDimension(doc.height);
  const nl = opts.prettyPrint ? '\n' : '';
  const ind = opts.prettyPrint ? '  ' : '';
  const parts: string[] = [];
  parts.push('<?xml version="1.0" encoding="UTF-8" standalone="no"?>');
  parts.push(
    `<svg xmlns="http://www.w3.org/2000/svg" version="1.1" ` +
      `width="${num(docW, p)}${escapeXml(opts.widthUnit)}" ` +
      `height="${num(docH, p)}${escapeXml(opts.widthUnit)}" ` +
      `viewBox="0 0 ${num(docW, p)} ${num(docH, p)}">`,
  );
  if (opts.includeMetadata) {
    parts.push(`${ind}<title>${escapeXml(opts.title)}</title>`);
    parts.push(
      `${ind}<desc>Generated by Offline Tracer. All processing performed on device.</desc>`,
    );
  }
  if (doc.background !== null) {
    const a = (doc.background >>> 24) & 0xff;
    const op = a !== 255 ? ` fill-opacity="${num(a / 255, 3)}"` : '';
    parts.push(
      `${ind}<rect x="0" y="0" width="${num(docW, p)}" height="${num(docH, p)}" ` +
        `fill="${colourHex(doc.background)}"${op}/>`,
    );
  }
  for (const layer of doc.layers) {
    if (opts.groupByLayer) {
      const attrs: string[] = [`id="${escapeXml(layer.id)}"`];
      attrs.push(`data-name="${escapeXml(layer.name)}"`);
      if (!layer.visible) attrs.push('display="none"');
      if (layer.opacity !== 1) attrs.push(`opacity="${num(layer.opacity, 3)}"`);
      parts.push(`${ind}<g ${attrs.join(' ')}>`);
      for (const shape of layer.shapes) parts.push(shapeElement(shape, opts, ind + ind));
      parts.push(`${ind}</g>`);
    } else {
      // Flattened output drops layer visibility, so a hidden layer must not silently become visible.
      if (!layer.visible) continue;
      for (const shape of layer.shapes) parts.push(shapeElement(shape, opts, ind));
    }
  }
  parts.push('</svg>');
  return parts.join(nl === '' ? '' : nl) + nl;
}

File: android/core-vector/src/main/java/com/offlinetracer/vector/SvgWriter.kt
package com.offlinetracer.vector

/**
 * SVG 1.1 serialisation of a [VecDocument].
 *
 * The output is deliberately plain: `width`/`height`/`viewBox` on the root, one `<g>` per layer
 * carrying the layer's id, name, opacity and visibility, and one `<path>` per shape. No transforms,
 * no `<defs>`, no CSS classes — every consumer of a traced file (Illustrator, Inkscape, a laser
 * cutter's job setup, a plotter, a browser) reads that subset identically, and each thing added
 * beyond it is another thing one of them reads differently.
 *
 * Two invariants matter more than the formatting:
 *
 *  - **Every attribute value is escaped.** Layer names come from the user, and an unescaped `&` in
 *    a layer name produces a file no XML parser will open.
 *  - **`NaN` and `Infinity` are never emitted.** A single non-finite coordinate makes a renderer
 *    drop the whole path, so [SvgPathData.num] clamps instead; see its KDoc.
 */
object SvgWriter {

    data class SvgOptions(
        val precision: Int = 2,
        val includeMetadata: Boolean = true,
        val groupByLayer: Boolean = true,
        val prettyPrint: Boolean = true,
        val widthUnit: String = "px",
        val title: String = "Offline Tracer export",
    )

    private const val MAX_DIMENSION = 1.0e7f

    /**
     * Serialises [doc] to a complete SVG 1.1 document, including the XML declaration.
     *
     * A non-positive or non-finite document size falls back to 1 unit so the file still opens; a
     * shape with no segments is skipped rather than emitted as an empty `<path>`. Paths carrying
     * `strokeWidths` are written as filled variable-width outlines, since SVG has no way to express
     * a stroke whose width varies along its length.
     */
    fun write(doc: VecDocument, options: SvgOptions = SvgOptions()): String {
        val p = if (options.precision < 0) 0 else if (options.precision > 6) 6 else options.precision
        val nl = if (options.prettyPrint) "\n" else ""
        val i1 = if (options.prettyPrint) "  " else ""
        val i2 = if (options.prettyPrint) "    " else ""

        val w = sanitizeDimension(doc.width)
        val h = sanitizeDimension(doc.height)
        val unit = escapeXml(options.widthUnit)

        val sb = StringBuilder(4096)
        sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>").append(nl)
        sb.append("<svg xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\"")
        sb.append(" width=\"").append(SvgPathData.num(w, p)).append(unit).append('"')
        sb.append(" height=\"").append(SvgPathData.num(h, p)).append(unit).append('"')
        sb.append(" viewBox=\"0 0 ").append(SvgPathData.num(w, p)).append(' ')
            .append(SvgPathData.num(h, p)).append("\">").append(nl)

        if (options.includeMetadata) {
            sb.append(i1).append("<title>").append(escapeXml(options.title)).append("</title>").append(nl)
            sb.append(i1).append("<desc>")
                .append(escapeXml("${doc.shapeCount()} shapes, ${doc.nodeCount()} nodes"))
                .append("</desc>").append(nl)
        }

        val bg = doc.background
        if (bg != null && ((bg ushr 24) and 0xFF) != 0) {
            sb.append(i1).append("<rect x=\"0\" y=\"0\" width=\"").append(SvgPathData.num(w, p))
                .append("\" height=\"").append(SvgPathData.num(h, p))
                .append("\" fill=\"").append(colorHex(bg)).append('"')
            appendAlpha(sb, "fill-opacity", bg, p)
            sb.append("/>").append(nl)
        }

        for (layer in doc.layers) {
            if (options.groupByLayer) {
                sb.append(i1).append("<g id=\"").append(escapeXml(layer.id)).append('"')
                sb.append(" data-name=\"").append(escapeXml(layer.name)).append('"')
                val opacity = clamp01(layer.opacity)
                if (opacity < 1f) sb.append(" opacity=\"").append(SvgPathData.num(opacity, 3)).append('"')
                if (!layer.visible) sb.append(" style=\"display:none\"")
                if (options.includeMetadata && layer.locked) sb.append(" data-locked=\"true\"")
                sb.append('>').append(nl)
                for (shape in layer.shapes) appendShape(sb, shape, i2, nl, p, 1f)
                sb.append(i1).append("</g>").append(nl)
            } else {
                // Flattened output has nowhere to record visibility, so a hidden layer is simply
                // absent and its opacity is folded into every shape it owned.
                if (!layer.visible) continue
                val opacity = clamp01(layer.opacity)
                if (opacity <= 0f) continue
                for (shape in layer.shapes) appendShape(sb, shape, i1, nl, p, opacity)
            }
        }

        sb.append("</svg>").append(nl)
        return sb.toString()
    }

    /**
     * Escapes the five XML entities and drops the control characters XML 1.0 cannot represent.
     * Safe for both attribute values and element text. Returns [s] unchanged when nothing needs it.
     */
    fun escapeXml(s: String): String {
        var needs = false
        for (c in s) {
            if (c == '&' || c == '<' || c == '>' || c == '"' || c == '\'' || c.code < 0x20) {
                needs = true
                break
            }
        }
        if (!needs) return s
        val sb = StringBuilder(s.length + 16)
        for (c in s) {
            when (c) {
                '&' -> sb.append("&amp;")
                '<' -> sb.append("&lt;")
                '>' -> sb.append("&gt;")
                '"' -> sb.append("&quot;")
                '\'' -> sb.append("&apos;")
                else -> if (c.code >= 0x20 || c == '\t' || c == '\n' || c == '\r') sb.append(c)
            }
        }
        return sb.toString()
    }

    // ---------------------------------------------------------------------------------------
    // Internals
    // ---------------------------------------------------------------------------------------

    private fun appendShape(
        sb: StringBuilder,
        shape: VecShape,
        indent: String,
        nl: String,
        precision: Int,
        opacityMultiplier: Float,
    ) {
        val path = shape.path
        if (path.segments.isEmpty()) return
        val style = shape.style
        val opacity = clamp01(style.opacity) * opacityMultiplier

        val widths = path.strokeWidths
        val stroke = style.stroke
        if (widths != null && widths.isNotEmpty() && stroke != null && style.strokeWidth > 0f) {
            val outline = StrokeStyle.variableWidthOutline(path, widths, style.cap)
            if (outline.segments.isEmpty()) return
            sb.append(indent).append("<path d=\"").append(SvgPathData.toD(outline, precision)).append('"')
            sb.append(" fill=\"").append(colorHex(stroke)).append('"')
            sb.append(" fill-rule=\"nonzero\"")
            appendAlpha(sb, "fill-opacity", stroke, precision)
            if (opacity < 1f) sb.append(" opacity=\"").append(SvgPathData.num(opacity, 3)).append('"')
            if (path.id.isNotEmpty()) sb.append(" id=\"").append(escapeXml(path.id)).append('"')
            sb.append("/>").append(nl)
            return
        }

        sb.append(indent).append("<path d=\"").append(SvgPathData.toD(path, precision)).append('"')
        val fill = style.fill
        if (fill == null) {
            sb.append(" fill=\"none\"")
        } else {
            sb.append(" fill=\"").append(colorHex(fill)).append('"')
            sb.append(" fill-rule=\"")
                .append(if (style.fillRule == FillRule.EVENODD) "evenodd" else "nonzero").append('"')
            appendAlpha(sb, "fill-opacity", fill, precision)
        }
        if (stroke != null && style.strokeWidth > 0f && style.strokeWidth.isFinite()) {
            sb.append(" stroke=\"").append(colorHex(stroke)).append('"')
            sb.append(" stroke-width=\"").append(SvgPathData.num(style.strokeWidth, precision)).append('"')
            sb.append(" stroke-linecap=\"").append(capName(style.cap)).append('"')
            sb.append(" stroke-linejoin=\"").append(joinName(style.join)).append('"')
            if (style.join == LineJoin.MITER) {
                val limit = if (style.miterLimit.isFinite() && style.miterLimit >= 1f) style.miterLimit else 4f
                sb.append(" stroke-miterlimit=\"").append(SvgPathData.num(limit, precision)).append('"')
            }
            appendAlpha(sb, "stroke-opacity", stroke, precision)
        }
        if (opacity < 1f) sb.append(" opacity=\"").append(SvgPathData.num(opacity, 3)).append('"')
        if (path.id.isNotEmpty()) sb.append(" id=\"").append(escapeXml(path.id)).append('"')
        sb.append("/>").append(nl)
    }

    private fun appendAlpha(sb: StringBuilder, attribute: String, argb: Int, precision: Int) {
        val a = (argb ushr 24) and 0xFF
        if (a >= 255) return
        val p = if (precision < 3) 3 else precision
        sb.append(' ').append(attribute).append("=\"").append(SvgPathData.num(a / 255f, p)).append('"')
    }

    private fun colorHex(argb: Int): String {
        val sb = StringBuilder(7)
        sb.append('#')
        appendHexByte(sb, (argb ushr 16) and 0xFF)
        appendHexByte(sb, (argb ushr 8) and 0xFF)
        appendHexByte(sb, argb and 0xFF)
        return sb.toString()
    }

    private fun appendHexByte(sb: StringBuilder, v: Int) {
        val digits = "0123456789ABCDEF"
        sb.append(digits[(v ushr 4) and 0xF])
        sb.append(digits[v and 0xF])
    }

    private fun capName(cap: LineCap): String = when (cap) {
        LineCap.BUTT -> "butt"
        LineCap.ROUND -> "round"
        LineCap.SQUARE -> "square"
    }

    private fun joinName(join: LineJoin): String = when (join) {
        LineJoin.MITER -> "miter"
        LineJoin.ROUND -> "round"
        LineJoin.BEVEL -> "bevel"
    }

    private fun sanitizeDimension(v: Float): Float {
        if (!v.isFinite() || v <= 0f) return 1f
        return if (v > MAX_DIMENSION) MAX_DIMENSION else v
    }

    private fun clamp01(v: Float): Float = if (!v.isFinite() || v < 0f) 0f else if (v > 1f) 1f else v
}

// Nested to match the API contract's grouping under `object SvgWriter`; aliased at file scope so
// `com.offlinetracer.vector.SvgOptions` resolves too. :core-export refers to it unqualified, and a
// type that only resolves one way is a compile error in somebody else's module.
typealias SvgOptions = SvgWriter.SvgOptions
